src/services/pure-di.ts

```diff
@@ -0,0 +1,63 @@
+import axios, { AxiosStatic } from 'axios';
+import { HttpDopplerLegacyClient, DopplerLegacyClient } from './doppler-legacy-client';
+import { OnlineSessionManager, SessionManager } from './session-manager';
+
+interface AppConfiguration {
+  dopplerLegacyUrl: string;
+  dopplerLegacyKeepAliveMilliseconds: number;
+}
+
+export interface AppContainer {
+  axiosStatic: AxiosStatic;
+  appConfiguration: AppConfiguration;
+  dopplerLegacyClient: DopplerLegacyClient;
+  sessionManager: SessionManager;
+}
+
+export class AppCompositionRoot implements AppContainer {
+  private readonly instances: Partial<AppContainer>;
+
+  constructor(predefinedInstances: Partial<AppContainer> = {}) {
+    this.instances = { ...predefinedInstances };
+  }
+
+  private singleton<N extends keyof AppContainer, T extends AppContainer[N]>(
+    name: N,
+    factory: () => T,
+  ): T {
+    if (!this.instances[name]) {
+      this.instances[name] = factory();
+    }
+    return this.instances[name] as T;
+  }
+
+  get axiosStatic() {
```

Comment:
I am using properties in order to have lazy initialization.

Another option could be using factory methods, for example: `getAxiosStatic()`, but it does not match _AppContainer_ interface:

```js
export interface AppContainer {
  axiosStatic: AxiosStatic;
  appConfiguration: AppConfiguration;
  dopplerLegacyClient: DopplerLegacyClient;
  sessionManager: SessionManager;
}
```

Another option could be using all as factories, the same for the interface: 


```js
export interface AppContainer {
  getAxiosStatic: () => AxiosStatic;
  getAppConfiguration: () => AppConfiguration;
  getDopplerLegacyClient: () => DopplerLegacyClient;
  getSessionManager: () => SessionManager;
}
```